src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+
+	// microTesla -> Gauss
+	_px4_mag.set_scale(0.01f);
+
+	return success;
+}
+
+int32_t BMM350::FixSign(uint32_t inval, int8_t num_bits)
+{
+	int32_t power = 1 << (num_bits - 1); // Calculate 2^(num_bits - 1)
+	int32_t retval = static_cast<int32_t>(inval);
+
+	if (retval >= power) {
+		retval -= (power << 1); // Equivalent to retval = retval - (power * 2)
+	}
+
+	return retval;
+}
+
+int8_t BMM350::ReadOutRawData(float *out_data)
+{
+	if (out_data == NULL) {
+		return -1;
+	}
+
+	float temp = 0.0;
+	struct BMM350::raw_mag_data raw_data = {0};
+	float lsb_to_ut_degc[4];
+
+	// --- Start read_uncomp_mag_temp_data
+	uint8_t mag_data[14] = {0};
+	uint8_t raw_reg[14] = {0};
+
+	uint32_t raw_mag_x, raw_mag_y, raw_mag_z, raw_temp;
+	uint8_t cmd = static_cast<uint8_t>(Register::DATAX_XLSB);
+
+	uint8_t res = transfer(&cmd, 1, (uint8_t *)&raw_reg, sizeof(raw_reg));
+
+	if (res != PX4_OK) {
+		return -1;
+	}
+
+	// Throwaway first two bytes
+	memcpy(mag_data, raw_reg + 2, sizeof(mag_data) - 2);
+
+	raw_mag_x = mag_data[0] + ((uint32_t)mag_data[1] << 8) + ((uint32_t)mag_data[2] << 16);
+	raw_mag_y = mag_data[3] + ((uint32_t)mag_data[4] << 8) + ((uint32_t)mag_data[5] << 16);
+	raw_mag_z = mag_data[6] + ((uint32_t)mag_data[7] << 8) + ((uint32_t)mag_data[8] << 16);
+	raw_temp = mag_data[9] + ((uint32_t)mag_data[10] << 8) + ((uint32_t)mag_data[11] << 16);
+
+	raw_data.raw_x = FixSign(raw_mag_x, 24);
+	raw_data.raw_y = FixSign(raw_mag_y, 24);
+	raw_data.raw_z = FixSign(raw_mag_z, 24);
+	raw_data.raw_t = FixSign(raw_temp, 24);
+	// --- End read_uncomp_mag_temp_data
+
+	// --- Start update_dafault_coefiecients
+	float bxy_sens, bz_sens, temp_sens, ina_xy_gain_trgt, ina_z_gain_trgt, adc_gain, lut_gain;
+	float power;
+
+	bxy_sens = 14.55f;
+	bz_sens = 9.0f;
+	temp_sens = 0.00204f;
+
+	ina_xy_gain_trgt = 19.46f;
+
+	ina_z_gain_trgt = 31.0;
+
+	adc_gain = 1 / 1.5f;
+	lut_gain = 0.714607238769531f;
+
+	power = (float)(1000000.0 / 1048576.0);
+
+	lsb_to_ut_degc[0] = (power / (bxy_sens * ina_xy_gain_trgt * adc_gain * lut_gain));
+	lsb_to_ut_degc[1] = (power / (bxy_sens * ina_xy_gain_trgt * adc_gain * lut_gain));
+	lsb_to_ut_degc[2] = (power / (bz_sens * ina_z_gain_trgt * adc_gain * lut_gain));
+	lsb_to_ut_degc[3] = 1 / (temp_sens * adc_gain * lut_gain * 1048576);
+	// --- End update_default_coeficients
+
+	// --- Start read_out_raw_data
+	out_data[0] = (float)raw_data.raw_x * lsb_to_ut_degc[0];
+	out_data[1] = (float)raw_data.raw_y * lsb_to_ut_degc[1];
+	out_data[2] = (float)raw_data.raw_z * lsb_to_ut_degc[2];
+	out_data[3] = (float)raw_data.raw_t * lsb_to_ut_degc[3];
+
+	// Adjust temperature
+	if (out_data[3] > 0.0f) {
+		temp = (float)(out_data[3] - (1 * 25.49f));
+
+	} else if (out_data[3] < 0.0f) {
+		temp = (float)(out_data[3] - (-1 * 25.49f));
+
+	} else {
+		temp = (float)(out_data[3]);
+	}
+
+	out_data[3] = temp;
+
+	return res;
+}
+
+int8_t BMM350::ReadOTPWord(uint8_t addr, uint16_t *lsb_msb)
+{
+	if (lsb_msb == NULL) {
```

Comment:
not necessary. the function is private within the driver and never called with NULL ptr...